tests/org.dataflowanalysis.analysis.tests/src/org/dataflowanalysis/analysis/tests/dfd/MicroSecEndTest.java

```diff
@@ -0,0 +1,296 @@
+            .put("jferrater", List.of(0, 2, 3, 5, 6, 7, 8, 9, 18))
+            .put("koushikkothagal", List.of(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 18))
+            .put("mdeket", List.of(5))
+            .put("mudigal-technologies", List.of(0, 2, 4, 5, 7, 8, 11, 18))
+            .put("rohitghatol", List.of(10, 12, 18))
+            .put("spring-petclinic", List.of(0, 2, 3, 5, 6, 7, 8, 9, 18))
+            .put("sqshq", List.of(0, 7, 8, 9, 10, 11, 12, 18))
+            .put("yidongnan", List.of(0, 2, 3, 4, 5, 6, 7, 8, 9, 18))
+            .build();
+
+    public DFDConfidentialityAnalysis buildAnalysis(String name) {
+        var DataFlowDiagramPath = name + ".dataflowdiagram";
+        var DataDictionaryPath = name + ".datadictionary";
+
+        return new DFDDataFlowAnalysisBuilder().standalone()
+                .modelProjectName(PROJECT_NAME)
+                .usePluginActivator(Activator.class)
+                .useDataFlowDiagram(DataFlowDiagramPath)
+                .useDataDictionary(DataDictionaryPath)
+                .build();
+    }
+
+    public void performAnalysis(String model, int variant, Set<Integer> violationsSet) {
+        var analysis = buildAnalysis(model);
+        analysis.initializeAnalysis();
+        var flowGraph = analysis.findFlowGraphs();
+        flowGraph.evaluate();
+        
+        //Sanity check if cycles get detected
+        if(model.endsWith("sqshq_11")) {
+            assertTrue(flowGraph.wasCyclic());
+        }
+
+        assertFalse(flowGraph.getTransposeFlowGraphs()
+                .isEmpty());
+
+        Map<Integer, List<AbstractTransposeFlowGraph>> existenceViolations = new HashMap<>();
+        for (var transposeFlowGraph : flowGraph.getTransposeFlowGraphs()) {
+
+            hasLoggingServer(transposeFlowGraph, existenceViolations);
+
+            hasSecretManager(transposeFlowGraph, existenceViolations);
+
+            for (var vertex : transposeFlowGraph.getVertices()) {
+
+                hasGateway(violationsSet, variant, vertex);
+
+                hasAuthenticatedRerquest(violationsSet, variant, vertex);
+
+                hasAuthorizedEntrypoint(violationsSet, variant, vertex);
+
+                hasTransformedEntryIdentity(violationsSet, variant, vertex);
+
+                hasTokenValidation(violationsSet, variant, vertex);
+
+                hasLoginAttemptsRegulation(violationsSet, variant, vertex);
+
+                hasEncryptedEntryConnection(violationsSet, variant, vertex);
+
+                hasEncrytedInternalConnection(violationsSet, variant, vertex);
+
+                hasLocalLogging(violationsSet, variant, vertex);
+
+                hasLogSanitization(violationsSet, variant, vertex);
+
+                hasMessageBroker(violationsSet, variant, vertex);
+
+            }
+        }
+        var numOfTransposeFlowGraphs = flowGraph.getTransposeFlowGraphs()
+                .size();
+        if (existenceViolations.get(9)
```

Comment:
This code (lines 107-120) is hard to understand without proper method names or additional comments

Reply:
Especially now that I read the methods that create the violations I second my earlier comment, this is hard to trace and should be encapsulated and properly documented. These are like second-order violations, that trigger after the first violations have been identified?

Reply:
Created method checkCrossTransposeFlowGraphViolations (maybe there is a better naming)
Also added comments will make it better to understand.

Reply:
Great, way better to understand now!